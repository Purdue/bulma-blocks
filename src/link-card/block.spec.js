// link card test
jest.setTimeout(30000)
import {
    createNewPost,
    enablePageDialogAccept,
    getEditedPostContent,
    insertBlock,
} from '@wordpress/e2e-test-utils';

import {
    clickElementByText,
    blockStartup,
    clickRadio,
    clickCheckbox
} from '../test-helpers'

const block = {blockTitle: 'Link Card', blockName: 'bulma-blocks/link-card'}

describe( `ðŸ”¬ ${block.blockTitle} Block`, () => {
    beforeAll( async () => {
        await enablePageDialogAccept();
    } );
    beforeEach( async () => {
        await createNewPost();
    } );

    test( 'ðŸ”¬ Block should be available.', async () => {
        await insertBlock( block.blockTitle )

        // tests that the block is properly inserted and matches the existing snapshot
        expect(await page.$(`[data-type="${block.blockName}"]`)).not.toBeNull()
        expect( await getEditedPostContent()).toMatchSnapshot()
    } )

    describe( 'ðŸ”¬ Block Editor Fields', () => {
        test( 'ðŸ”Ž Button should open Media Library.', async () => {
            await blockStartup(block)

            // open media library
            await clickElementByText('button', 'Open Media Library')

            // tests that media library does open
            expect(await page.$(`div.media-frame-title`)).not.toBeNull()

            /*
            NOTE: explore further to see how media items can be spoofed so that the actual
                image urls can be properly snapshotted and tested.
            */
        })
        
        test( 'ðŸ”Ž Card Text Field', async () => {
            await blockStartup(block)

            const typeString = "Test Card Text"
            
            // focus the text box then type into it with the virtual keyboard
            await page.focus( '.input[placeholder="Card text..."]' );
            await page.keyboard.type(typeString, {delay: 10})

            const editedContent = await getEditedPostContent()

            expect( editedContent).toMatchSnapshot()
            expect( editedContent.includes(`"subText":"${typeString}"`)).toBe(true)
        })

        test( 'ðŸ”Ž Card Link Text', async () => {
            await blockStartup(block)

            const typeString = "Test Card Link Text"
            
            // focus the text box then type into it with the virtual keyboard
            await page.focus( '.input[placeholder="Link text..."]' );
            await page.keyboard.type(typeString, {delay: 10})

            const editedContent = await getEditedPostContent()

            expect( editedContent).toMatchSnapshot()
            expect( editedContent.includes(`"linkText":"${typeString}"`)).toBe(true)
        })

        test( 'ðŸ”Ž Card Link URL', async () => {
            await blockStartup(block)

            const typeString = "Test Card Link Url"
            
            // focus the text box then type into it with the virtual keyboard
            await page.focus( '.input[placeholder="Paste permalink or url..."]' );
            await page.keyboard.type(typeString, {delay: 10})

            const editedContent = await getEditedPostContent()

            expect( editedContent).toMatchSnapshot()
            expect( editedContent.includes(`"link":"${typeString}"`)).toBe(true)
        })
    })
    describe( 'ðŸ”¬ Side Panel Settings', () => {
        test( 'ðŸ”Ž Card Layout Radio Options.', async () => {
            await blockStartup(block)

            // horizontal
            await clickRadio('Layout of the card', 'horizontal')

            let editedContent = await getEditedPostContent()

            expect( editedContent.includes(`"layout":"vertical"`)).toBe(false)
            expect( editedContent).toMatchSnapshot()

            // vertical
            await clickRadio('Layout of the card', 'vertical')

            editedContent = await getEditedPostContent()

            expect( editedContent.includes(`"layout":"vertical"`)).toBe(true)
            expect( editedContent).toMatchSnapshot()
        })
        test( 'ðŸ”Ž Image alt text should be editable.', async () => {
            await blockStartup(block)

            const typeString = "Image alt text."
            
            // focus the text box then type into it with the virtual keyboard
            await clickElementByText('label', 'Image Alt Text')
            await page.keyboard.type(typeString, {delay: 10})

            const editedContent = await getEditedPostContent()

            expect( editedContent).toMatchSnapshot()
            expect( editedContent.includes(`"altText":"${typeString}"`)).toBe(true)
        })
        test( 'ðŸ”Ž Open link in new tab checkbox', async () => {
            await blockStartup(block)
            
            await clickCheckbox('Open link in new tab?')

            const editedContent = await getEditedPostContent()

            expect( editedContent).toMatchSnapshot()
            expect( editedContent.includes(`"external":true`)).toBe(true)
        })
    })

} );
