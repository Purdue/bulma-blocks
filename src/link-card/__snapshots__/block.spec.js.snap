// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`🔬 Link Card Block 🔬 Block Editor Fields 🔎 Card Link Text 1`] = `
"<!-- wp:bulma-blocks/link-card {\\"linkText\\":\\"Test Card Link Text\\"} -->
<a href=\\"\\" target=\\"_self\\" class=\\"card media link-card-horizontal\\" rel=\\"noopener noreferrer\\"><div class=\\"media-left\\"><figure class=\\"image\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"media-content\\"><p class=\\"title is-4\\"></p><div class=\\"read-more-button\\"><span>Test Card Link Text</span></div></div></a>
<!-- /wp:bulma-blocks/link-card -->"
`;

exports[`🔬 Link Card Block 🔬 Block Editor Fields 🔎 Card Link URL 1`] = `
"<!-- wp:bulma-blocks/link-card {\\"link\\":\\"Test Card Link Url\\"} -->
<a href=\\"Test Card Link Url\\" target=\\"_self\\" class=\\"card media link-card-horizontal\\" rel=\\"noopener noreferrer\\"><div class=\\"media-left\\"><figure class=\\"image\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"media-content\\"><p class=\\"title is-4\\"></p><div class=\\"read-more-button\\"><span>Read More</span></div></div></a>
<!-- /wp:bulma-blocks/link-card -->"
`;

exports[`🔬 Link Card Block 🔬 Block Editor Fields 🔎 Card Text Field 1`] = `
"<!-- wp:bulma-blocks/link-card {\\"subText\\":\\"Test Card Text\\"} -->
<a href=\\"\\" target=\\"_self\\" class=\\"card media link-card-horizontal\\" rel=\\"noopener noreferrer\\"><div class=\\"media-left\\"><figure class=\\"image\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"media-content\\"><p class=\\"title is-4\\">Test Card Text</p><div class=\\"read-more-button\\"><span>Read More</span></div></div></a>
<!-- /wp:bulma-blocks/link-card -->"
`;

exports[`🔬 Link Card Block 🔬 Block should be available. 1`] = `
"<!-- wp:bulma-blocks/link-card -->
<a href=\\"\\" target=\\"_self\\" class=\\"card media link-card-horizontal\\" rel=\\"noopener noreferrer\\"><div class=\\"media-left\\"><figure class=\\"image\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"media-content\\"><p class=\\"title is-4\\"></p><div class=\\"read-more-button\\"><span>Read More</span></div></div></a>
<!-- /wp:bulma-blocks/link-card -->"
`;

exports[`🔬 Link Card Block 🔬 Side Panel Settings 🔎 Card Layout Radio Options. 1`] = `
"<!-- wp:bulma-blocks/link-card -->
<a href=\\"\\" target=\\"_self\\" class=\\"card media link-card-horizontal\\" rel=\\"noopener noreferrer\\"><div class=\\"media-left\\"><figure class=\\"image\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"media-content\\"><p class=\\"title is-4\\"></p><div class=\\"read-more-button\\"><span>Read More</span></div></div></a>
<!-- /wp:bulma-blocks/link-card -->"
`;

exports[`🔬 Link Card Block 🔬 Side Panel Settings 🔎 Card Layout Radio Options. 2`] = `
"<!-- wp:bulma-blocks/link-card {\\"layout\\":\\"vertical\\"} -->
<a href=\\"\\" target=\\"_self\\" class=\\"card media link-card\\" rel=\\"noopener noreferrer\\"><div class=\\"image is-2by1 background-image\\" role=\\"img\\" style=\\"background-image:url()\\" aria-label=\\"\\"></div><div class=\\"media-content\\"><p class=\\"title is-4\\"></p></div></a>
<!-- /wp:bulma-blocks/link-card -->"
`;

exports[`🔬 Link Card Block 🔬 Side Panel Settings 🔎 Image alt text should be editable. 1`] = `
"<!-- wp:bulma-blocks/link-card {\\"altText\\":\\"Image alt text.\\"} -->
<a href=\\"\\" target=\\"_self\\" class=\\"card media link-card-horizontal\\" rel=\\"noopener noreferrer\\"><div class=\\"media-left\\"><figure class=\\"image\\"><img src=\\"\\" alt=\\"Image alt text.\\"/></figure></div><div class=\\"media-content\\"><p class=\\"title is-4\\"></p><div class=\\"read-more-button\\"><span>Read More</span></div></div></a>
<!-- /wp:bulma-blocks/link-card -->"
`;

exports[`🔬 Link Card Block 🔬 Side Panel Settings 🔎 Open link in new tab checkbox 1`] = `
"<!-- wp:bulma-blocks/link-card {\\"external\\":true} -->
<a href=\\"\\" target=\\"_blank\\" class=\\"card media link-card-horizontal\\" rel=\\"noopener noreferrer\\"><div class=\\"media-left\\"><figure class=\\"image\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"media-content\\"><p class=\\"title is-4\\"></p><div class=\\"read-more-button\\"><span>Read More</span></div></div></a>
<!-- /wp:bulma-blocks/link-card -->"
`;
